pennylane/qchem/factorization.py

```diff
@@ -135,46 +214,192 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
         - Diagonalize the :math:`n \times n` matrices and for each matrix keep the eigenvalues (and
           their corresponding eigenvectors) that are larger than a threshold.
     """
-    shape = two_electron.shape
+    shape = qml.math.shape(two_electron)
 
     if len(shape) != 4 or len(set(shape)) != 1:
         raise ValueError("The two-electron repulsion tensor must have a (N x N x N x N) shape.")
 
-    n = shape[0]
-    two = two_electron.reshape(n * n, n * n)
+    two = qml.math.reshape(two_electron, (shape[0] * shape[1], -1))
+    interface = qml.math.get_interface(two_electron)
+
+    if not compressed:
+        if cholesky:
+            factors = _double_factorization_cholesky(two, tol_factor, shape, interface)
+        else:
+            factors = _double_factorization_eigen(two, tol_factor, shape, interface)
+
+        core_tensors, leaf_tensors = [], []
+        feigvals, feigvecs = qml.math.linalg.eigh(factors)
+        for feigval, feigvec in zip(feigvals, feigvecs):
+            fidx = qml.math.where(qml.math.abs(feigval) > tol_eigval)[0]
+            core_tensors.append(qml.math.einsum("i,j->ij", feigval[fidx], feigval[fidx]))
+            leaf_tensors.append(feigvec[:, fidx])
+
+        if np.sum([len(v) for v in core_tensors]) == 0:
+            raise ValueError(
+                "All eigenvectors are discarded. Consider decreasing the second threshold error."
+            )
+
+    else:
+        if not has_jax_optax:
+            raise ImportError(
+                "Jax and Optax libraries are required for optimizing the factors. Install them via "
+                "pip install jax optax"
+            )  # pragma: no cover
+
+        norm_order = {None: None, "L1": 1, "L2": 2}.get(regularization, None)
+        optimizer = compression_kwargs.get("optimizer", optax.adam(learning_rate=0.001))
+        num_steps = compression_kwargs.get("num_steps", 1000)
+        num_factors = compression_kwargs.get("num_factors", 2 * shape[0])
+        prefactor = compression_kwargs.get("norm_prefactor", 1e-5)
+        init_params = compression_kwargs.get("init_params", None)
+
+        if cholesky and init_params is None:
+            factors = _double_factorization_cholesky(two, tol_factor, shape, interface, num_factors)
+            f_vals, f_vecs = qml.math.linalg.eigh(factors)
+            core_matrices = qml.math.einsum("ti,tj->tij", f_vals, f_vals)
+            leaf_matrices = [sp.linalg.logm(vec).real for vec in f_vecs]
+            num_factors = qml.math.shape(core_matrices)[0]
+            init_params = {"X": leaf_matrices, "Z": core_matrices}
```

Comment:
Can you bring this inside _double_factorization_cholesky? Or put a comment explaining what it is doing.